File: backend/src/driver/main_test.go
// Copyright 2025 The Kubeflow Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"testing"

	"github.com/kubeflow/pipelines/backend/src/driver/driverapi"
	"github.com/kubeflow/pipelines/backend/src/v2/driver"
	"github.com/kubeflow/pipelines/kubernetes_platform/go/kubernetesplatform"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"google.golang.org/protobuf/proto"
)

func strPtr(s string) *string {
	return &s
}

func TestSpecParsing(t *testing.T) {
	tt := []struct {
		name     string
		input    *string
		expected *kubernetesplatform.KubernetesExecutorConfig
		wantErr  bool
	}{
		{
			"Valid - test kubecfg value parse.",
			strPtr("{\"imagePullSecret\":[{\"secret_name\":\"value1\"}]}"),
			&kubernetesplatform.KubernetesExecutorConfig{
				ImagePullSecret: []*kubernetesplatform.ImagePullSecret{
					{SecretName: "value1"},
				},
			},
			false,
		},
		{
			"Valid - test kubecfg value ignores unknown field.",
			strPtr("{\"imagePullSecret\":[{\"secret_name\":\"value1\"}], \"unknown_field\": \"something\"}"),
			&kubernetesplatform.KubernetesExecutorConfig{
				ImagePullSecret: []*kubernetesplatform.ImagePullSecret{
					{SecretName: "value1"},
				},
			},
			false,
		},
	}

	for _, tc := range tt {
		t.Logf("Running test case: %s", tc.name)
		cfg, err := parseExecConfigJSON(tc.input)
		assert.Equal(t, tc.wantErr, err != nil)
		assert.True(t, proto.Equal(tc.expected, cfg))
	}
}

func Test_extractOutputParametersContainer(t *testing.T) {
	execution := &driver.Execution{}

	outputs := extractOutputParameters(execution, CONTAINER)

	require.NotEmpty(t, outputs)
	verifyOutputParameter(t, outputs, "condition", "nil")
}

func Test_extractOutputParametersRootDAG(t *testing.T) {
	execution := &driver.Execution{}

	outputs := extractOutputParameters(execution, RootDag)
	require.NotEmpty(t, outputs)
	verifyOutputParameter(t, outputs, "iteration-count", "0")
	verifyOutputParameter(t, outputs, "condition", "nil")
}

func Test_extractOutputParametersDAG(t *testing.T) {
	execution := &driver.Execution{}

	outputs := extractOutputParameters(execution, DAG)

	require.NotEmpty(t, outputs)
	verifyOutputParameter(t, outputs, "iteration-count", "0")
	verifyOutputParameter(t, outputs, "condition", "nil")
}

func verifyOutputParameter(t *testing.T, parameters []driverapi.Parameter, key, expectedValue string) {
	filtered := make([]driverapi.Parameter, 0, 1)
	for _, p := range parameters {
		if p.Name == key {
			filtered = append(filtered, p)
		}
	}
	require.Len(t, filtered, 1)
	require.Equal(t, expectedValue, filtered[0].Value)
}

File: sdk/python/kfp/cli/__main__test.py
# Copyright 2024 The Kubeflow Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tests for kfp.cli.__main__."""

import sys
import unittest
from unittest import mock

from kfp.cli import __main__


class TestMain(unittest.TestCase):

    @mock.patch('kfp.cli.__main__.cli.cli')
    def test_main_success(self, mock_cli):
        __main__.main()
        mock_cli.assert_called_once_with(obj={}, auto_envvar_prefix='KFP')

    @mock.patch('kfp.cli.__main__.cli.cli')
    @mock.patch('kfp.cli.__main__.click.echo')
    @mock.patch('kfp.cli.__main__.sys.exit')
    def test_main_exception(self, mock_exit, mock_echo, mock_cli):
        mock_cli.side_effect = Exception('test error')
        __main__.main()
        mock_echo.assert_called_once_with('test error', err=True)
        mock_exit.assert_called_once_with(1)

    @mock.patch('kfp.cli.__main__.logging.basicConfig')
    def test_main_logging_config(self, mock_basic_config):
        with mock.patch('kfp.cli.__main__.cli.cli'):
            __main__.main()
        mock_basic_config.assert_called_once_with(
            format='%(message)s', level=20)


if __name__ == '__main__':
    unittest.main()
